---
title: "Differential PA site usage analysis"
author: "Huayun Hou"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
 bookdown::html_document2:
    toc: true
    toc_float: true
    code_folding: hide
---

This script performs differential PA site usage analysis using R package *DEXSeq*.

# Script set up and read in data

```{r setup, include=TRUE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, eval = FALSE)
# source essential functions
# source("~/Dropbox/collab_brent/batch2/scripts/derfinder_ce11_utr_functions.r")
genename <- readRDS("gdata/Caenorhabditis_elegans.WBcel235.85.genenames.rds")

library(DEXSeq)
library(openxlsx)
library(tidyverse)
library(patchwork)
library(gplots)
library(ggrepel)


#load("01_PA_site_identification.RData")
# read in count table prepared for DEXSeq analysis
dexcount_df <-  readRDS("cele_all_utrcounts_dexcounts_df.rds")
sampleCondition <- readRDS("ce11_sampleCondition.rds")
sampleCondition$condition <- factor(sampleCondition$condition, levels =  c("N2", "daf2", "daf16", "cfim1", "daf2-cfim1","daf2-daf16"))

load("02_differential_PA_usage.RData")
```


```{r addfuncs}
# Additional functions used 
dexseq_run <- function(compare_pair, sampleTable=sampleCondition){
  # function to run dexseq pairwise comparisons
  print(compare_pair)
  compare_pair <- as.character(compare_pair)
  used <- which(sampleTable$condition %in% compare_pair)
  count_table_used <- dexcount_df[,used]
  sampleTable_used <- sampleTable[used,]
  dxd = DEXSeqDataSet(
    count_table_used,
    sampleData=sampleTable_used,
    design= ~ sample + exon + condition:exon,
    groupID = sapply(strsplit(row.names(count_table_used), ":"), "[[",1),
    featureID = sapply(strsplit(row.names(count_table_used), ":"), "[[",2))
  
  dxd <- estimateSizeFactors(dxd)
  dxd <- estimateDispersions(dxd)
  
  dxd <- testForDEU(dxd)
  dxd <- estimateExonFoldChanges(dxd, fitExpToVar="condition")
  
  result <- DEXSeqResults(dxd)
  result <- result[order(result$padj),]
  result <- as.data.frame(result)
  result$genename <- genename[result$groupID]
  return(result)
}

reformat_dexresult <- function(sig_list_used, comparison){
  print(comparison)
  test <- sig_list_used[[comparison]]
  # add region intervals
  test$genomicData <- paste(named_regions[as.character(row.names(test))])
  # find out which samples are used for this comparison
  cons <- unlist(strsplit(comparison, "_vs_"))
  samples_used <- as.character(subset(sampleCondition, condition %in% cons)$sample)
  # rename countData columns with samplenames
  colnames(test)[grepl("countData", colnames(test))] <- samples_used
  test <- test[!duplicated(test[,c("genomicData", "genename")]),]
  return(test)
}

assign_direction <- function(comparison){
  sig_df <- sig_list[[comparison]]
  compare_groups <- unlist(strsplit(comparison, "_vs_"))
  #compare_groups <- gsub("log2fold_", "", names(sig_df)[10])
  #compare_groups <- unlist(strsplit(compare_groups, "_"))
  directions <- ifelse(sig_df[, grepl("fold", colnames(sig_df))] >0, compare_groups[1], compare_groups[2])
  output <- paste0(compare_groups[1], "_vs_", compare_groups[2], ":", directions)
  return(data.frame(geneid=sapply(strsplit(rownames(sig_df),":"),"[[",1), 
                   exon=sapply(strsplit(rownames(sig_df),":"),"[[",2),
                   bias=output,
                   uniqid=rownames(sig_df),
                   stringsAsFactors = F))
}

label_dexseqresults <- function(comparison, change_label =F){
  dexresult_sum <- assign_direction(comparison)
  dexresult_sum$bias <- sapply(strsplit(dexresult_sum$bias, ":"), "[[", 2)
  if(change_label){
    dexresult_sum$bias[dexresult_sum$bias != "N2"] <- "mut"
  }
  dexresult_sum$exon <- ifelse(dexresult_sum$exon == "E001", "proximal", "distal")
  con1 <- unlist(strsplit(comparison, "_vs_"))[[1]]
  con2 <- unlist(strsplit(comparison, "_vs_"))[[1]]
  out <- dexresult_sum %>% 
    mutate(condition=paste(exon, bias, sep="_")) %>%
    group_by(geneid) %>%
    summarise(APAresult=paste(sort(unique(condition)), collapse=":")) %>%
    as.data.frame()
  topcons <- names(sort(-table(out$APAresult)))[1:4]
  out$APAresult_simple <- sapply(out$APAresult, function(x) ifelse(x %in% topcons,x,"others"))
  out$genename <- genename[as.character(out$geneid)]
  return(out)
}

```

# Perform differential PA site usage comparisons between any pair of conditions

```{r dexseq_run}
compare_pairs <- combn(unique(sampleCondition$condition), 2)
compare_pairs <- lapply(1:ncol(compare_pairs), function(x) sort(as.character(compare_pairs[,x])))
names(compare_pairs) <- sapply(compare_pairs, function(x) paste(as.character(x), collapse="_vs_"))


result_list <- mclapply(compare_pairs, dexseq_run, sampleCondition, mc.cores = 4)
names(result_list) <- names(compare_pairs)
#saveRDS(result_list, paste0("ce11_DEXseq_results_all_", Sys.Date(), ".rds"))

```

## select for significant differentially used PA sites

We used a cut off of adjusted p value < 0.05 between strains to capture significant PA usage changes regardless of the extand of change. 

```{r sig}
# fix comparison names 
new_names <- gsub(".","-",gsub("_","_vs_", unname(gsub("log2fold_","",sapply(result_list, function(x) names(x)[grepl("log2", names(x))])))), fixed = T)
used_comparisons <- lapply(new_names, function(x) unlist(strsplit(x, "_vs_")))
#saveRDS(used_comparisons, paste0("ce11_DEXseq_comparison_", Sys.Date(), ".rds"))

# filter for sites with 
sig_list <- lapply(result_list, function(x) subset(x,padj < 0.05))
names(sig_list) <- new_names

sig_list_output <- lapply(names(sig_list), function(x) reformat_dexresult(sig_list, x))
names(sig_list_output) <- names(sig_list)

# output reformatted table, without fold change cutoff
write.xlsx(sig_list_output, "cele_utrcounts_DEXSeq_results_pairwise.xlsx")
```


# Summarize differential PA site usage per gene

PA site usage changes (APA events) are classified for each gene. The most proximal PA site is denoted as "proximal" while the rest are denoted "distal". Thus, there could be multiple distal sites. 

For each PA site, it is labeled as **proximal/distal**_**condition in which its more highly used**. Thus, a site labeled as "proximal_cfim1" would mean that compared to N2, this proximal PA site is more highly used in cfim1.


```{r label, eval=T}
comparisons <- c("cfim1_vs_N2","daf2-cfim1_vs_N2")
APAresult_sum <- lapply(comparisons, label_dexseqresults, change_label = F)
names(APAresult_sum) <- comparisons
#saveRDS(APAresult_sum, paste0("ce11_APAresult_sum.rds"))

#write.xlsx(APAresult_sum,paste0("summary_of_APAresults_allsamples.xlsx"))

APAresult_sum_used <- lapply(comparisons, label_dexseqresults, change_label = T)
names(APAresult_sum_used) <- comparisons
APAresult_sum_df <- bind_rows(APAresult_sum_used, .id = "id")

```

# Generating plots to summarize 

We only chose to summarize the occurance of four simple cases and labeled the rest as "others". 

Bar plot summarizes the occurance of each case in the comparisons we chose. 

In the heatmap, each row is a gene, color represents the type of APA events for that gene in different comparisons. If the gene did not have any significant PA site usage change identified, it will be colored grey. 

```{r plot, eval=TRUE, fig.cap=c("barplot showing APA event type summary for selected comparisons", "heatmap showing per gene APA event type in selected comparisons")}
# examine the occurance of different combinations
# APAresult_sum_df %>%
#   group_by(APAresult, L1) %>%
#   summarize(n=length(geneid)) %>%
#   arrange(-n)

show <- c("distal_N2:proximal_mut","distal_mut:proximal_N2","proximal_mut","distal_N2")

APAresult_sum_df$label <- APAresult_sum_df$APAresult
APAresult_sum_df$label[!APAresult_sum_df$label %in% show] <- "others"

APAresult_sum_df <- APAresult_sum_df %>%
  mutate(label=factor(label, levels = rev(c(show, "others"))), id = factor(id, levels=c("cfim1_vs_N2", "daf2-cfim1_vs_N2")))

# pairedcols <- colorRampPalette(brewer.pal(5, 'Paired'))(5)
# plot gene APA cases as bars 
ggplot(APAresult_sum_df) +
  geom_bar(aes(x=id, fill=label), position="stack") +
 # scale_fill_manual(values = rev(pairedcols))+
  scale_fill_manual(values = comparison_cols) +
  theme_bw() +
  xlab("comparisons") +
  ylab("number of genes with APA") +
  scale_y_continuous(expand=c(0,0)) +
    theme(axis.text.x = element_text(angle=30, hjust = 1, color="black"),
          legend.title = element_blank())

#ggsave(paste0("APAresult_summary_labels_selected.pdf"), scale=0.8, width=6, height = 4.5)


# plot gene APA cases as a heatmap
# Adding NA values to cases without matched genes
temprank <- spread(APAresult_sum_df[,c("geneid", "id","label")], id, label) %>%
  arrange(`cfim1_vs_N2`,`daf2-cfim1_vs_N2`) %>%
  as.data.frame()

temp <- APAresult_sum_df[, c("geneid", "label", "id")] %>%
  #filter(label %in% c("cfim1_vs_N2","daf2-cfim1_vs_N2")) %>% 
  spread(geneid, label, fill=NA) %>%
  gather(geneid, label, -id) %>%
  mutate(geneid=factor(geneid, levels=rev(as.character(temprank$geneid))))

# plot
ggplot(temp) +
  geom_tile(aes(x=id, y=geneid, fill=factor(label, levels = comparison_labels))) +
  scale_fill_manual(values = comparison_cols, na.value="grey60")+
  theme_bw()+
  theme(panel.background=element_blank(),
        axis.title=element_blank(),
        panel.grid=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks=element_blank(),
        axis.text.x=element_text(angle=30, color="black",hjust=1),
        legend.title = element_blank())
#ggsave("APAresult_summary_heatmap_of_labels_selected.pdf", scale=0.8, width=5, height = 8)

```

```{r countOL, eval=T, fig.cap="Number of genes with higher usage of the proximal PA site in mutants"}

cfim1_vs_N2 <- subset(APAresult_sum_df, id == "cfim1_vs_N2" & grepl("proximal_mut", APAresult))$geneid
daf2_cfim1_vs_N2 <- subset(APAresult_sum_df, id == "daf2-cfim1_vs_N2" & grepl("proximal_mut", APAresult))$geneid

plot(venn(list("cfim1_vs_N2" =cfim1_vs_N2, "daf2-cfim1_vs_N2" = daf2_cfim1_vs_N2)))

```


# plots showing PA site read counts per gene

```{r generate-gene-plots, eval=T, fig.cap=c("barplot showing normalized read counts for each PA site of selected genes"), fig.width=10}

# generate bar plots showing number of reads for a certain gene
# load count table
get_sem <- function(x){
  return(round(sd(x)/sqrt(length(x)),2))
}

count_table_file <- list.files(pattern = glob2rx("ce11*count_table*rds"))
count_table <- readRDS(count_table_file)
count_sum <- colSums(count_table)
# normalize to CPM
norm_dexcounts <- round(as.data.frame(t(t(dexcount_df)/(count_sum/1e6))), 2)

colnames(norm_dexcounts) <- sampleCondition$sample 
plot_gene <- "mrp-5"
plot_cons <- c("N2", "cfim1", "daf2-cfim1")

plot_PAsite_bar <- function(plot_gene, plot_cons = c("N2", "cfim1", "daf2-cfim1")){
  plot_geneid <- names(genename[which(genename == plot_gene)])
  
  plot_samples <- as.character(subset(sampleCondition, condition %in% plot_cons)$sample)
  
  gene_counts <- norm_dexcounts %>%
    rownames_to_column("info") %>%
    separate(info, sep = ":", into = c("gene_id", "PAsite")) %>%
    filter(gene_id == plot_geneid) %>%
    gather(sample, counts, -gene_id, -PAsite) %>%
    filter(sample %in% plot_samples) %>%
    separate(sample, sep = "_", into = c("id", "condition", "rep")) %>%
    mutate(PAsite = recode(PAsite, "E001" = "proximal", "E002" = "distal"), 
           condition = factor(condition, levels = plot_cons)) %>% 
    mutate(PAsite = factor(PAsite, levels = c("proximal", "distal")))
  
  gene_counts_mean <- gene_counts %>%
    group_by(gene_id, PAsite, condition) %>%
    summarise(mean = mean(counts), sem = get_sem(counts)) 
  
  p_counts_bar <- ggplot() +
    geom_bar(data = gene_counts_mean, aes(x = PAsite, y = mean, fill = condition), stat = "identity", position = position_dodge(.8), width = .7) +
    geom_errorbar(data = gene_counts_mean, aes(x = PAsite, ymin = mean-sem, ymax = mean+sem, group = condition, color = condition), position = position_dodge(.8), width = .3) +
  #   geom_point(data = gene_counts, aes(x = PAsite, y = counts, color = condition), alpha = .7, position= position_jitterdodge()) +
    scale_fill_manual(values = condition_col_list$condition) +
    scale_color_manual(values = condition_col_list$condition) +
    scale_y_continuous(expand = c(0,0))+
    #geom_point(data = gene_counts, aes(x = PAsite, y = counts, color = condition)) +
    theme_classic() +
    theme(axis.text.x = element_text(color = "black", angle = 30, hjust = 1, vjust = 1, size = 12),
          axis.text.y = element_text(color = "black", size = 12),
          axis.title = element_text(size = 14)) +
    xlab(paste(plot_gene, " isoforms")) +
    ylab("CPM (counts per million)")
  
  #ggsave(paste0(plot_gene, "_PAsite_counts_bar.pdf"), p_counts_bar, width = 4, height = 3)
  return(list(gene_counts, p_counts_bar))
}

gene_counts_list <- lapply(c("mrp-5","pak-1", "ftt-2"),  plot_PAsite_bar)
#write.table(do.call("rbind", lapply(gene_counts_list, "[[",1)), "gene_plot_counts.txt", col.names = T, quote = F, row.names = F, sep = "\t")

gene_counts_list[[1]][[2]]+gene_counts_list[[2]][[2]] +gene_counts_list[[3]][[2]] +
  plot_layout(guides = "collect")

```

# gene expression vs. PPAU

```{r expr-vs-ppau, eval=T}
# change condition "-" to "_"
row.names(sampleCondition) <- colnames(count_table)
sampleCondition$condition2 <- gsub("-","_",sampleCondition$condition)
sampleCondition$sample <- as.character(sampleCondition$sample)

# use DEseq2 for count normalization
dds <- DESeqDataSetFromMatrix(countData = count_table, 
                              colData = sampleCondition,
                              design = ~ condition2)

dds <- DESeq(dds)
logCPM <- log2(fpm(dds, robust =TRUE) + 1)

logCPM_out <- as.data.frame(logCPM)
logCPM_out$geneID <- row.names(logCPM_out)
logCPM_out$geneName <- genename[logCPM_out$geneID]
#write.xlsx(logCPM_out, "cele_allsamples_logCPM_DEseq2_normed.xlsx")

# load ppau results
ppau_file <- list.files(pattern = glob2rx("cele_all_utrindex_df*rds"))
ppau_df <- readRDS(ppau_file)
ppau_df <- ppau_df[, -c(1:2)]

ppau_diff_vs_cpm <- function(comparison = "cfim1_vs_N2", logCPM, ppau_df){
  con1 <- strsplit(comparison, "_vs_")[[1]][1]
  con2 <- strsplit(comparison, "_vs_")[[1]][2]
  ppau_1 <- ppau_df[, subset(sampleCondition, condition == con1)$sample]
  ppau_2 <- ppau_df[, subset(sampleCondition, condition == con2)$sample]
  ppau_diff <- data.frame(gene_id = row.names(ppau_df),
                          ppau_diff = rowMeans(ppau_1) - rowMeans(ppau_2),
                          stringsAsFactors = F)  

  logCPM_mean <- data.frame(gene_id = row.names(logCPM),
                            mean_logCPM = rowMeans(logCPM[, subset(sampleCondition, condition %in% c(con1, con2))$sample]),
                            stringsAsFactors = F)

  merged <- left_join(ppau_diff, logCPM_mean, by = "gene_id") %>% 
    left_join(subset(APAresult_sum_df, id == comparison), by = c("gene_id" = "geneid")) %>% 
    mutate(ifsig = ifelse(is.na(id),"no","yes"))
  
  p_violin <- ggplot(merged, aes(x = 1, y = ppau_diff)) +
    geom_violin() +
    geom_boxplot(width = .2) +
  #  geom_hline(yintercept = 0, linetype = 2) +
    ylab(bquote(~pPAU[mut]~' - '~pPAU[WT])) +
    theme_linedraw() +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          axis.text.y = element_blank(),
          axis.title.y = element_text(size = 12, angle = 90),
          axis.ticks.x = element_blank(),
          axis.ticks.y = element_blank())
  
  p <- ggplot(merged) +
    geom_point(aes(x = mean_logCPM, y = ppau_diff, color = ifsig), alpha = .5) +
    geom_hline(yintercept = 0, linetype = 2) +
    theme_linedraw() +
    scale_color_manual(values = c("yes" = "tomato", "no" = "grey60")) +
    xlab("mean log2 (CPM)") +
    ylab(bquote(~pPAU[mut]~' - '~pPAU[WT])) +
    ggtitle(comparison) +
    theme(axis.text = element_text(size = 12),
          axis.title.x = element_text(size = 12),
          axis.title.y = element_blank(),
          title = element_text(size = 12),
          panel.grid = element_blank()) 
  
  p2 <- p_violin +p + plot_layout(widths = c(1,6))
  
  return(list(p2, merged))
}

ppau_cpm_cfim1_N2 <- ppau_diff_vs_cpm(comparison = "cfim1_vs_N2", ppau_df = ppau_df, logCPM = logCPM)
ppau_cpm_cfim1_N2[[1]]
#ggsave(ppau_cpm_cfim1_N2[[1]], filename = "cfim1_vs_N2_pPAU_diff_vs_logCPM.pdf", width = 6, height = 5,useDingbats = F)

ppau_cpm_daf2cfim1_N2 <- ppau_diff_vs_cpm(comparison = "daf2-cfim1_vs_N2", ppau_df = ppau_df, logCPM = logCPM)
ppau_cpm_daf2cfim1_N2[[1]]

#ggsave(ppau_cpm_daf2cfim1_N2[[1]], filename = "daf2-cfim1_vs_N2_pPAU_diff_vs_logCPM.pdf", width = 6, height = 5,useDingbats = F)

# save ppau df
ppau_out <- as.data.frame(ppau_df) %>% 
  rownames_to_column("gene_id") %>% 
  mutate(gene_name = genename[gene_id]) %>% 
  select(gene_id, gene_name, everything())
#write.xlsx(ppau_out, "cele_allsamples_ppau.xlsx")
```


# Example gene plots

## Supplementry Figure 1C
```{r suppfig1c, eval=T}

plot_genes <- c("cfim-1", "mtl-1", "sod-3", "dod-6")
plot_con <- c("N2", "daf2", "daf2-daf16")
plot_samples <- as.character(subset(sampleCondition, condition %in% plot_con)$sample)

logCPM_genename <- as.data.frame(logCPM) %>% 
  rownames_to_column("geneID") %>% 
  mutate(geneName = genename[geneID])

used_data <- logCPM_genename %>% 
  filter(geneName %in% plot_genes) %>% 
  select(plot_samples, geneName, -geneID) %>% 
  gather(sample,value, -geneName) %>% 
  left_join(sampleCondition, by = "sample") %>% 
  mutate(geneName = factor(geneName, levels = c("cfim-1","mtl-1","sod-3","dod-6")))

used_data_sum <- used_data %>% 
  group_by(geneName, condition) %>% 
  summarise(mean = mean(value), sem = get_sem(value)) 

ggplot() +
  geom_col(data = used_data_sum, aes(x = geneName, y = mean, color = condition), fill = NA,position = "dodge") +
  geom_point(data = used_data, aes(x = geneName, y = value, color = condition), alpha = .7, position= position_jitterdodge()) +
   geom_errorbar(data = used_data_sum, aes(x = geneName, ymin = mean - sem, ymax = mean + sem, color = condition), position = position_dodge(width = .8), width = .5) +
  scale_color_manual(values = c("black","orange","blue")) +
  theme_classic() +
  scale_y_continuous(expand = c(0,0), name = "log2 normalized counts") +
  theme(axis.text.x = element_text(size = 16, color = "black",  angle = 30, hjust = 1, vjust = 1),
        axis.text.y = element_text(size = 16, color = "black"),
          axis.title.y = element_text(size = 16),
         axis.title.x = element_blank())
  
#ggsave("supp_fig1C_example_genes_logCPM_v2.pdf", width = 5, height = 3)


ggplot() +
  geom_col(data = used_data_sum, aes(x = geneName, y = mean, fill = condition),position = "dodge", width = .8) +
  geom_errorbar(data = used_data_sum, aes(x = geneName, ymin = mean - sem, ymax = mean + sem, fill = condition), color = "black", position = position_dodge(width = .8), width = .5) +
 #  geom_point(data = used_data, aes(x = geneName, y = value, fill = condition), alpha = .7, position= position_jitterdodge()) +
  scale_fill_manual(values = c("black","orange","blue")) +
  theme_classic() +
  scale_y_continuous(expand = c(0,0), name = "log2 normalized counts") +
  theme(axis.text.x = element_text(size = 16, color = "black",  angle = 30, hjust = 1, vjust = 1),
        axis.text.y = element_text(size = 16, color = "black"),
          axis.title.y = element_text(size = 16),
         axis.title.x = element_blank())
#ggsave("supp_fig1C_example_genes_logCPM_v1.pdf", width = 5, height = 3)
```

## Supplementary figure 4b

```{r suppfig4b, eval=T}

plot_genes <- c("egl-20","kgb-1","T20H4.5","rab-6.2","C05D9.3","pyp-1","pak-1","R119.3","sem-5","top-1","capg-1","usp-46")

con1 <- "daf2-cfim1"
con2 <- "N2"

ppau_df_used <- data.frame(ppau1 = rowMeans(ppau_df[, as.character(subset(sampleCondition, condition == con1)$sample)]),
                      ppau2 = rowMeans(ppau_df[, as.character(subset(sampleCondition, condition == con2)$sample)]))
ppau_df_used$geneName <- genename[row.names(ppau_df_used)]

logcpm_df_used <- data.frame(cpm1 = rowMeans(logCPM[, subset(sampleCondition, condition == con1)$sample]),
                      cpm2 = rowMeans(logCPM[, subset(sampleCondition, condition == con2)$sample]))
logcpm_df_used$geneName <- genename[row.names(logcpm_df_used)]

merged <- logcpm_df_used %>% 
  filter(geneName %in% plot_genes) %>% 
  left_join(subset(ppau_df_used, geneName %in% plot_genes), by = "geneName") %>% 
  mutate(diffUTR = ppau1-ppau2, diffcpm=cpm1 - cpm2)

R2 <- summary(lm(diffUTR ~ diffcpm, merged))$r.squared
pvalue <- summary(lm(diffUTR ~ diffcpm, merged))$coefficients[2,4] 

ggplot(merged, aes(x = diffUTR, y = diffcpm)) +
  geom_smooth(method = "lm", se = FALSE, color = "red") +
  geom_point(color = "red", size = 4) +
  geom_text_repel(aes(label = geneName), size = 6) +
  theme_classic() +
  annotate(geom = "text", label = paste0("R^2=", round(R2,2),"\n","p=", round(pvalue,2)), x = .5, y = 3, size = 6) +
  xlab("UTR index (daf-2;cfim-1 - N2)") +
  ylab("mRNA: log2 normalized counts\n (daf-2;cfim-1/N2)") +
  theme(axis.text = element_text(size = 16, color = "black"),
          axis.title = element_text(size = 16)) 
  
#ggsave("supp_4b_ppau_vs_cpm_difference_example_genes.pdf", width = 6, height = 5, useDingbats = F)

```



```{r save}
save.image("02_differential_PA_usage.RData")
```

```{r sessioninfo, message=T, eval=T, warning=T}
sessionInfo()
```

